### ISSUE TITLE
Constrained dynamics

### ISSUE THREAD

**Issue by nickgerrits** [i-pi/i-pi#221]
Dear all,

I've been working on constrained dynamics with i-Pi and I've noticed some peculiar things and bug(s).

I've tried running some test simulations involving the constraints "distance" and "angle" but they don't seem to work. This seems to be a bug in inputs/motion/constrained_dynamics.py:
135: "alist.shape = (alist.shape[0] / 2, 2)"
145: "alist.shape = (alist.shape[0] / 3, 3)"

Changing this to
135: "alist.shape = (int(alist.shape[0] / 2), 2)"
145: "alist.shape = (int(alist.shape[0] / 3), 3)"
seems to resolve the problem. I'm not sure whether this is caused by the use of particular Python versions, but in my case this worked.

Another more worrying thing is the fact that within constrained dynamics the atomic positions within each bead don't seem to be updated, only the centroid and first bead. For example, when dumping the atomic position of an atom within each individual bead through printing atom_x or atom_x_path in properties shows clearly that only the positions of the centroid and first bead are changing. Also, I've attached LAMMPS as a driver and let it dump all positions so I could check whether positions were actually changing or not. And indeed, only the position of the first bead changes, whereas the positions of all the other beads remain constant.

Weirdly enough, the momenta are adjusted for all beads (e.g., use atom_v in properties) but I'm not sure whether they are correct. I've also implemented my own constraint where atomic positions can be fixed, which works somewhat similar to fixatoms. (Note that fixatoms and fixcom are not implemented within constrained dynamics, hence the need for a fixatom constraint.) When you apply such a fix you would expect momenta to be close to zero (since you are solving the constraint numerically, you might deviate from zero due to numerical noise), but I definitely see much larger values than that, even when positions are not changing.

Either I'm really confused on what constrained dynamics is trying to achieve, or something is really going wrong. In its current form it basically only updates the positions of a single bead and it seems to not touch the positions of the other beads. This also results in the constrained dynamics to yield something completely different compared to classical (i.e., single bead) simulations: The centroid is an average of the static non-updated beads and the single bead that is correctly (?) following the constrained dynamics.

Finally, I've noticed there is another numerical problem with the constrained dynamics. Right now, the constraint solver relies on Cholesky decomposition of the relevant matrices to obtain the correction to the positions and momenta when scipy is used (which, by the way, occurs in engine/motion/constrained_dynamics.py and indeed seems to be only updating the first bead and comparison with engine/motion/dynamics.py suggests to me that the implementation is simply incorrect or incomplete). However, this then also needs the matrix to be solved to be positive definite, and there is simply no guarantee that this will be the case during constrained dynamics and the various constraints. In fact, when I implemented my own constraints I quickly found that Cholesky decomposition regularly lets the program crash, and ironically even more when the system is very close to the constraint, i.e., the correction is numerically extremely close to or exactly zero. Changing the solver to scipy's linalg lstsq function seems to be much more robust and might be preferable in the future. That, or first trying Cholesky (or LU) decomposition combined with some error catching and have lstsq as back-up.

I'm looking forward to your response,

Nick Gerrits
Postdoctoral researcher at PLASMANT - University of Antwerp

**Comment by ceriottm** [comment]
Hi Nick. Unfortunately constraints have not been used much by us core developers, so it's a bit tricky for us to help. the `alist.shape` bugs are likely due to the conversion from python2 to python 3, the right way would be to have e.g. `alist.shape = (alist.shape[0] // 2, 2)` i.e. use an integer division. @MatthiasSachs  did some work on the implementation of constraints, and perhaps has enough bandwidth to chime in and help you clean up the implementation - which should be pretty much complete, although not fully tested.

**Comment by nickgerrits** [comment]
Hi Michele. Yes, sorry, I haven't read PEP8 yet, so I'm not always aware what the most "Pythonic" way is. Also thanks for referring it to Matthias.

The problem is that I think the implementation is not complete, especially after further examination of dynamics vs constrained_dynamics. In normal dynamics, you first convert the beads from cartesian to normal mode coordinates. Then, the position of the first normal mode (i.e., the centroid) is moved with regular velocity verlet, whereas the rest of the normal modes is moved with an appropriate RPMD propagator. However, in the constrained dynamics no normal mode transformation is performed and only the momenta and positions of the first bead are touched (really, I couldn't find a single instance in that part of the code that touches one of the other beads). That means constrained dynamics in its current form should only work in the classical limit (i.e., a single bead). I think the only part of the constrained dynamics that works somewhat is the solver (with the caveat of the instability of the Cholesky decomposition and the fact that it calculates the constraint's value on the first bead instead of the centroid) and the fact that the code can adjust the number of dofs.

However, even the latter is done in a not so ideal way. It is assumed that the number of constraints employed is related to the number of dofs that are being removed. That means that if you implement an efficient constraint that removes multiple dofs simultaneously (e.g., fix an atom completely in XYZ), you need to divide up the constraint in several pieces for the solver. This complicates solving a constraint even further. Normally, you would rather have a single constraint, let it solve (can often also be done exactly, instead of numerically as is done here) for the solution, and have a definition in the constraint that determines the number of dofs removed. See for example how ASE has implemented it.

In any case, I would argue that the constrained dynamics in the current form is completely unusable and a big warning needs to be included for others to not use it (I really wasted a lot of time before I found the single bead problem). Furthermore, it might require an almost full rewrite.

**Comment by MatthiasSachs** [comment]
Dear both, 
apologies for my late reply! I was traveling during the past days with only limited access to my computer. 

Regarding the implementation and some of the issues you are raising: as far as I remember (@ceriottm might want to correct me) our implementation indeed was intended to only apply constraints to a classical system, thus the issues you are mentioning. There were some extensions added to the code after the first implementation was finished that include constraints related to the the center of mass (as far as I remember; I didn't follow this as closely as I should have).

I am a bit constrained time-wise for the next two weeks, but would nonetheless be happy to help cleaning-up the current implementation. To my understanding this would mainly involve extending the current implementation to multiple beats and possibly fixing issues related to the computation of the matrix square root. I believe, that a Cholesky decomposition with pivoting would be sufficient. 

I suggest to continue further discussion via email. Feel free to email me on the email address provided here on GitHub! 

Best,
Matthias

**Comment by litman90** [comment]
Dear all,

If you don't mind, keeping the conversation on GitHub would be great. 
This way, Michele, I, or other interested developers might participate as well. Also, the rationale behind some design decisions won't get lost and can be traced back/discussed in the future if required.
We can continue here or one can open a discussion in the Discussion section
All the best,
Yair

**Comment by ceriottm** [comment]
I agree, let's keep the discussion on github. I will open a branch in
i-pi-cosmo and a draft pull request - we can start with an error message
when nbeads>1, and then see how far we get. I would set as a minimal goal
to make sure that constraints are reliable for classical trajectories, and
having centroid constraints as a bonus.

**Comment by ceriottm** [comment]
Done. Let's move the discussion to the PR #224. You should all be able to push to https://github.com/lab-cosmo/i-pi/tree/feat/constraints_reloaded

**Comment by ceriottm** [comment]
Hi Nick, I believe that the constrained dynamics was implemented only for
classical MD - and indeed there should be a warning, or even better an
error raised when constraints are applied with nbeads>1.
As far as I know, doing constrained dynamics with PIMD is far from trivial
- from a conceptual perspective even more than from a practical
implementation viewpoint. Defining the quantum mechanical partition
function for a constrained problem is tricky - I vaguely recall a paper
from the '90s where they did PIMD for rigid water and it is really a mess.
One approach would be to apply the constraint to the centroid, but even
there it's not clear to me what would be the physical underpinning, and the
practical goal.
Can you explain better what is it you are trying to achieve? I suspect that
within the scope of applicability of traditional constrained MD (that is,
classical trajectories) this is not far from being a working
implementation, Matthias tested it quite thoroughly for liquid water, back
then.

**Comment by mahrossi** [comment]
Hello, should we close this or do you want to still keep this open? @ceriottm - we did some work on constraints but not general ones.

**Comment by ceriottm** [comment]
I think that the issue is not solved and it'd be nice for constraints to be implemented properly. I would leave it open so that if one of us ever needs constraints they may find the motivation to clean up properly rather than do a half-*ssed hack.
